feat: add h2c sample for Cloud Run

ahmetb opened the pull request:
I'm contributing a pure HTTP2 cleartext (h2c) server example as part
of HTTP/2 launch for Cloud Run.

This server doesn't support upgrading from HTTP/1 intentionally to
show that end-to-end HTTP/2 works on Cloud Run.

I intentionally have not added tests as the tests mainly would be
testing x/net/http2 package, which is not relevant to test here,
and would make the code more complicated (to make it testable).

cc: @averikitsch @grayside

ahmetb commented:
Some errors it seems:

```
=== RUN   TestBadFiles
    badfiles_test.go:147: Likely bad file checked in: run/h2c/app.json. MIME type: {  }
```

`app.json` will be needed for sample app to be deployable via `cloud-run-button` (to specify --use-http2).

```
=== RUN   TestLicense
    license_test.go:75: run/h2c/main.go: license header not present
```

This license was added with our own https://github.com/google/addlicense tool. I'm attaching license for debugging:

<details>
<summary>main.go</summary>
<pre>
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
</pre>
</details>

averikitsch commented:
Can you add a system test that builds and deploys this? I agree that unit tests aren't necessary here

ahmetb commented:
Added system tests.
I had to introduce service.{HTTP2,AsBuildpack} fields. As a result some gcloud commands now use "beta" or "alpha" prefix. I added TODOs to address those later.

ahmetb commented:
Tests seem stuck, any ideas how to re-test?
Also I think this is now ready to review again.

ahmetb commented:
I get some integration test fails from `spanner` that are not related to what I updated:

- https://source.cloud.google.com/results/invocations/26e36a06-4bb3-4c55-94d1-892cab5549cb/targets/github%2Fgolang-samples/tests
- https://source.cloud.google.com/results/invocations/26e36a06-4bb3-4c55-94d1-892cab5549cb/targets/cloud-devrel%2Fgo%2Fgolang-samples%2Fpresubmit%2Fgo111/log

grayside commented on run/h2c/README.md:
````diff
@@ -0,0 +1,59 @@
+    go run .
+    ```
+
+## Verifying h2c
````
In general we don't have README's unless the particular sample requires it, in preference to maintaining instructions only in the official docs. I'm not sure if this section would be added to the docs, so I'm in favor of keeping this README and keeping it here at least until the scope of HTTP2 content in the Cloud Run docs is settled.

grayside commented on run/h2c/main.go:
```diff
@@ -0,0 +1,55 @@
+// Copyright 2020 Google LLC
```
New sample should specify 2021

grayside commented on run/h2c/main.go:
```diff
@@ -0,0 +1,55 @@
+	"golang.org/x/net/http2"
+)
+
+func main() {
```
nit: Do we think side-by-side comparison with the "usual" way we start the server is important? If so there are a few syntactic differences that could be more aligned with the other samples.

ahmetb replied:
I don't think there's much value in going into too much detail. Likely people who know Go will refer to this, as every language will be vastly different in terms of how to support h2c.

grayside commented on internal/cloudrunci/cloudrunci.go:
```diff
@@ -259,6 +265,7 @@ func (s *Service) operationLabel(op string) string {
 func (s *Service) deployCmd() *exec.Cmd {
 	args := append([]string{
 		"--quiet",
+		"alpha", // TODO until --use-http2 goes GA
```
I didn't realize the alpha component was installed, nor do I see it installed in the code, but since their are other, not-skipped tests that use `gcloud alpha` it must be fine.

ahmetb replied:
I suspect --quiet accepts installation if it's not installed.

grayside reviewed (approved):
I left a couple comments that may be worth action before merging, but nothing that needs to block.

ahmetb commented:
Fixed the year.
I kept the README.md because similar sample `run/grpc-server-streaming/` also has it, also this one shows the user how to test deployment locally and on Cloud Run.
